### PR TITLE
Verify the block filter hash when reading from disk.

### PR DESCRIPTION
**Pull request by pstratem**
What's stored in the block filter index data files is the block_hash and the encoded gcs filter.

If there's a disk error that corrupts the encoded filter there's no way for the node to know.

Clients that use the "header" chain will notice the error though.

It takes 1ms to verify the hash on a filter with 10000 elements.

### CODE REVIEW

**Comment by pstratem** [comment]
the Travis failure doesn't seem related

**Review by jonatack — COMMENTED** [review]
Code review ACK 180e094d52262322d525ccdef9, did not benchmark the performance cost.

**Comment by MarcoFalke** [comment]
We don't do this when reading blocks from disk and serving them to p2p. Also, with silent corruption on disk, wouldn't it be better to abort the node?

**Comment by pstratem** [comment]
@MarcoFalke that's a good point, I've changed it so that we don't run the checks in GCSFilter::GCSFilter when the encoded filter hash has been checked.  That check is about as expensive as calculating the hash.

**Comment by pstratem** [comment]
I've added the benchmarks I used.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 75e86b18846aa21a88463dfd3e710534afb6c104. If I'm understanding this correctly, it is replacing slower less comprehensive checks that filters contain the right number of elements with faster checks verifying complete hashes of the filters. It also adds a benchmark. So this seems better in every respect.

File: src/index/blockfilterindex.cpp [hunk 1]
```diff
@@ -153,7 +154,10 @@ bool BlockFilterIndex::ReadFilterFromDisk(const FlatFilePos& pos, BlockFilter& f
     std::vector<unsigned char> encoded_filter;
     try {
         filein >> block_hash >> encoded_filter;
-        filter = BlockFilter(GetFilterType(), block_hash, std::move(encoded_filter));
+        uint256 result;
+        CHash256().Write(encoded_filter.data(), encoded_filter.size()).Finalize(result.begin());
```
Comment (jonatack) [inline src/index/blockfilterindex.cpp:158]: This looks like a future candidate to become a `Span`.

**Review by jonatack — COMMENTED** [review]
ACK 75e86b1 per `git diff 180e094 75e86b1`

Ran comparative benchmarks built with `--enable-bench CXXFLAGS="-O2"`.

This branch rebased on master:
```
# Benchmark, evals, iterations, total, min, max, median
DecodeCheckedGCSFilter, 100, 50000, 4.28914, 7.30889e-07, 1.02493e-06, 8.58921e-07

DecodeGCSFilter, 10, 5000, 16.2784, 0.000314371, 0.000337642, 0.000329746

BlockFilterGetHash, 10, 10000, 15.4826, 0.000142116, 0.000163642, 0.000158849

ConstructGCSFilter, 10, 1000, 22.988, 0.00217746, 0.00253763, 0.00227277

MatchGCSFilter, 10, 50000, 18.8786, 3.56681e-05, 4.12871e-05, 3.7422e-05
```

Master, with the standardized BASIC filter params. Ran twice:
```
# Benchmark, evals, iterations, total, min, max, median
ConstructGCSFilter, 10, 1000, 23.2998, 0.00226351, 0.00248678, 0.00233295
ConstructGCSFilter, 10, 1000, 23.2406, 0.00225238, 0.00251339, 0.00230733

MatchGCSFilter, 10, 50000, 19.3379, 3.70859e-05, 4.26215e-05, 3.85519e-05
MatchGCSFilter, 10, 50000, 18.7122, 3.54355e-05, 3.90555e-05, 3.78147e-05
```

**Comment by MarcoFalke** [comment]
This had two acks before conflicting. @pstratem are you still working on this or should someone pick this up?

**Pull request closed without merging**